core/Access.php

```diff
@@ -689,6 +699,28 @@ public function getCapabilitiesForSite($idSite)
         }
         return $result;
     }
+
+    /**
+     * Throw a NoAccessException with the given message, or a more generic 'You need to log in' message if the
+     * user is not currently logged in (e.g. if session has expired).
+     * @param $message
+     * @throws NoAccessException
+     */
+    private function throwNoAccessException($message)
+    {
+        if (Piwik::isUserIsAnonymous()) {
+            $message = Piwik::translate('General_YouMustBeLoggedIn');
+        }
+        // Try to detect whether user was previously logged in so that we can display a different message
+        $referrer = Url::getReferrer();
+        if (Url::isValidHost(Url::getHostFromUrl($referrer)) &&
```

Comment:
@katebutler could we check also that `referrer` is not empty? Not sure if needed but may be good.